plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/xtext/server/XProjectManager.java

```diff
@@ -128,18 +131,37 @@ public void initialize(ProjectDescription description, IProjectConfig projectCon
 		this.resourceSet = createNewResourceSet(new XIndexState().getResourceDescriptions());
 	}
 
-	/** Initial build reads the project state and resolves changes. Generate output files. */
+	/**
+	 * Initial build reads the project state and resolves changes. Generate output files.
+	 * <p>
+	 * NOTE: this is not only invoked shortly after server startup but also at various other occasions, for example
+	 * <ul>
+	 * <li>when the client executes the {@link N4JSCommandService#rebuild(ILanguageServerAccess, CancelIndicator)
+	 * rebuild command},
+	 * <li>when the workspace folder is changed in VS Code.
+	 * </ul>
+	 */
 	public XBuildResult doInitialBuild(CancelIndicator cancelIndicator) {
 		ResourceChangeSet changeSet = projectStateHolder.readProjectState(projectConfig);
 		XBuildResult result = doBuild(
 				changeSet.getModified(), changeSet.getDeleted(), Collections.emptyList(), false, true, cancelIndicator);
 
+		// send issues to client
 		Map<URI, Collection<Issue>> validationIssues = projectStateHolder.getValidationIssues();
+		Set<URI> urisWithoutIssues = result.getAffectedResources().stream()
+				.map(IResourceDescription.Delta::getUri)
+				.collect(Collectors.toSet());
 		for (Map.Entry<URI, Collection<Issue>> locationToIssues : validationIssues.entrySet()) {
 			URI location = locationToIssues.getKey();
+			urisWithoutIssues.remove(location);
 			Collection<Issue> issues = locationToIssues.getValue();
 			issueAcceptor.publishDiagnostics(location, issues);
 		}
+		// send empty 'publishDiagnostics' for resources without issues, even deleted ones
+		// (to remove potential obsolete issues)
+		for (URI uri : urisWithoutIssues) {
+			issueAcceptor.publishDiagnostics(uri, Collections.emptyList());
```

Comment:
Why would we have stale issues on the client for the initial build?

Reply:
1) This is also invoked from the command "N4JS: Rebuild" (as I had noted in the API doc a few lines further up).
2) Also, it seemed to me that the LSP spec is not 100% clear that is is forbidden for clients to serialize and store/restore issues; so if some client other than VS Code would do that, this would even be required for a build right after server startup.

Reply:
:+1:

Shame on me that I didn't spot this in the JavaDoc. Point taken ;)